sdk-api-src/content/winusb/nf-winusb-winusb_readisochpipe.md

```diff
@@ -106,10 +106,11 @@ Pointer to an <a href="https://docs.microsoft.com/windows/desktop/api/shobjidl/n
 
 
 
-<b>WinUsb_ReadIsochPipe</b> packetizes the transfer buffer so that in each interval,  the host can receive the maximum bytes allowed per interval. The maximum bytes is as specified by the endpoint descriptor for full and high-speed endpoints, and endpoint companion descriptor for SuperSpeed endpoints.
+<b>WinUsb_ReadIsochPipe</b> packetizes the transfer buffer so that in each 1ms interval,  the host can receive the maximum bytes allowed per interval. The maximum bytes is as specified by the endpoint descriptor for full and high-speed endpoints, and endpoint companion descriptor for SuperSpeed endpoints.
 If the caller submits multiple read requests to stream data from the device, the transfer size should be a multiple of the maximum bytes per interval (as returned by <a href="https://docs.microsoft.com/windows/desktop/api/winusb/nf-winusb-winusb_querypipeex">WinUsb_QueryPipeEx</a>) * 8 / interval.
 
 
+Because of the transfer packaging, <b>only full 1ms USB frames are available</b> and low-latency applications requiring microframe transfers must use a kernel mode driver instead.
```

Comment:
I'm not sure I understand this statement.  Regardless of whether WinUsb or a kernel-mode driver is used, the underlying kernel-mode interface only allows isochronous transfers to be scheduled on a 1ms frame boundary.  For high-speed/SuperSpeed devices which microframes are scheduled is determined by the interval of the endpoint.  

Whether a developer needs to use a kernel-mode driver versus WinUSB for low-latency applications is determined by the latency between completion of the isoch transfer on the bus and when the driver or app is notified of that completion.  It has nothing to do with frames versus micro-frames.

Reply:
I agree with both statements. But the critical point here is that lowest latency notification of the driver/app is only done at the 1ms interval regardless of the bInterval of the endpoint. How about:

```suggestion
Because of the transfer packaging used in the underlying kernel-mode interface, the lowest latency notification to an application or driver is 1ms intervals. This also applies to high-speed/SuperSpeed devices regardless of the interval specified by the endpoint. Applications requiring sub-millisecond latency must implement a custom kernel-mode driver.
```

Reply:
I would only take the typo fixes for these changes.  Discussion about the need for a user-mode app versus kernel-mode driver should not be part of the API documentation.

Reply:
Maybe, but details about the imposed limitations of the implementation is definitely useful for anyone using the API in latency critical applications.